src/benchmarklib/benchmark_runner.cpp

```diff
@@ -267,20 +274,23 @@ void BenchmarkRunner::_warmup_query(const QueryID query_id) {
 }
 
 std::vector<std::shared_ptr<AbstractTask>> BenchmarkRunner::_schedule_or_execute_query(
-    const QueryID query_id, const std::function<void()>& done_callback) {
+    const QueryID query_id, const bool prepare, const std::function<void()>& done_callback) {
   // Some queries (like TPC-H 15) require execution before we can call get_tasks() on the pipeline.
   // These queries can't be scheduled yet, therefore we fall back to "just" executing the query
   // when we don't use the scheduler anyway, so that they can be executed.
   if (_config.enable_scheduler) {
-    return _schedule_query(query_id, done_callback);
+    return _schedule_query(query_id, prepare, done_callback);
   }
-  _execute_query(query_id, done_callback);
+  _execute_query(query_id, prepare, done_callback);
   return {};
 }
 
 std::vector<std::shared_ptr<AbstractTask>> BenchmarkRunner::_schedule_query(
-    const QueryID query_id, const std::function<void()>& done_callback) {
-  const auto& sql = _query_generator->build_query(query_id);
+    const QueryID query_id, const bool prepare, const std::function<void()>& done_callback) {
+  const auto& sql =
+      prepare ? _query_generator->preparation_queries()[query_id] : _query_generator->build_query(query_id);
+
+  if (prepare && sql.empty()) return {};
```

Comment:
sql can be empty? Does this imply that every BenchmarkQueryGeneratorThing needs to provide an empty PreparationQuery for each of its queries?

Reply:
Yes. For example, TPC-H 15 cannot be prepared.

The alternative would be a method `bool query_needs_preparation`, but I don't see that improving the code. What we could also do is have a single, large SQL string that prepares all queries for a given benchmark.

Reply:
That options sounds somewhat attractive to me. Maybe `AbstractQueryGenerator::get_setup_query()` or similar?

Reply:
("preparation_query"; I know its derived from PREPARE; but I didn't immediately associate it with "ah, this returns PREPARE statements")

Reply:
Mmmh, if we go down that road, shouldn't the query generator simply prepare everything as part of its constructor? I don't see an advantage in that back-and-forth.

Reply:
Could do that, but I don't like constructors doing work (because there MIGHT be virtual function calls hidden in that work, you never know) and I think there is some value in the BenchmarkRunner logging something along the lines of "- Running setup query: PREPARE a, PREPARE b, ...` and clearly informing the user what is happening.

Reply:
To throw in my opinion here as well: I like having separate SQL strings better than one big one to be honest. And if you just split the `_execute` method and have a separate `_prepare` method as you suggested in the other thread, you can just replace the `_execute_query(query, true)` in the beginning with `_prepare_query(query)` and we're good, right?
This gets rid of the boolean parameter in the schedule and and execute methods, keeps the preparation in the BenchmarkRunner to inform the user, and is very clear to the user.